Build tsh with static libfido2 for macOS

codingllama opened the pull request:
Add a script to build libfido2 (and its dependencies) on macOS and enable FIDO2 static builds.

I decided to build all dependencies instead of pulling from Homebrew for a few reasons:

1. There is no libcbor.a in a brew package
2. This captures library versions within the Teleport source code, allowing us to build binaries against different versions of libfido2 (and its dependencies).

I've also bumped libfido2 to 1.11.0. I've been running it locally and we are still pre-release, so it seems like a good time to do it. (See https://developers.yubico.com/libfido2/Release_Notes.html.)

#9160

codingllama committed:
Build libfido2 and dependencies for macOS

codingllama committed:
Build tsh with static fido2 on Drone

codingllama committed:
Regenerate Drone files

codingllama committed:
Bump libfido2 versions in all builds

codingllama commented on .drone.yml:
```diff
@@ -534,8 +534,10 @@ steps:
   - export RUST_HOME=$CARGO_HOME
   - export PATH=~/build-$DRONE_BUILD_NUMBER-$DRONE_BUILD_CREATED-toolchains/go/bin:$CARGO_HOME/bin:/Users/build/.cargo/bin:$PATH
   - cd $WORKSPACE_DIR/go/src/github.com/gravitational/teleport
+  - build.assets/build-fido2.sh build
```
Note to reviewers: this file is autogenerated.

codingllama commented:
This should repro the builds if you are on macOS:

```shell
build.assets/build-fido2.sh build
export PKG_CONFIG_PATH="$(build.assets/build-fido2.sh pkg_config_path)"
pkg-config libfido2-static  # returns 0
make build/tsh FIDO2=yes
# GOOS=darwin GOARCH=amd64 CGO_ENABLED=1 go build -tags " libfido2 libfido2static " -o build/tsh  -ldflags '-w -s' ./tool/tsh
# ^- notice libfido2 and libfido2static build tags
otool -L build/tsh    # no libfido2, libcbor or libcrypto
build/tsh fido2 diag  # works
```

FYI @zmb3

codingllama committed:
Attempt to appease linters

espadolini commented on build.assets/build-fido2.sh:
```diff
@@ -0,0 +1,209 @@
+  mkdir -p "$LIB_CACHE"
+  if [[ ! -d "$dest" ]]; then
+    echo "$name: moving $fullname to $dest" >&2
+    mv "$install_path" "$dest"
```
Isn't this still racy - in the sense of erroring out if a race condition happens?

codingllama replied:
Yep, there's a small change this will still fail due to a race condition. I don't think we have enough concurrent builds for it to be a serious problem, though.

codingllama replied:
Adding a bit more detail, I chose to have it fail in the off-chance of a race instead of swallowing a possible mv error (say, `mv src dest || true`), which risks moving forward in a possible bad state. This should only happen if there is a race in the "fist batch" of building a new version, and retrying the failed builder would have it work (without having to rebuild anything, so not much time is lost). As said above, I think the chance for flakiness here is rare, so I didn't try to be clever about it.

espadolini commented on build.assets/build-fido2.sh:
```diff
@@ -0,0 +1,209 @@
+  trap "rm -fr '$tmp'" exit
+
+  local fullname="$name-$version"
+  local install_path="$tmp/$fullname"
```
Have you considered having `install_path` be a tmpdir in `LIB_CACHE` so the final install is atomic?

codingllama committed:
Use temp dirs inside LIB_CACHE

codingllama committed:
Move LIB_CACHE outside of HOME

HOME is reassigned in macOS builders, but we want a "stable" cache
directory. /tmp is used by build-package.sh and build-pkg-tsh.sh.

codingllama commented:
PTAL?

espadolini commented on build.assets/build-fido2.sh:
```diff
@@ -0,0 +1,209 @@
+# Run `build-fido2.sh pkg_config_path` to print the path to the prior-built
+# libfido2-static.pc file.
+#
+# Written mainly for macOS builders.
```
Should the filename of the script reflect that this is very much only for macOS?

codingllama replied:
Sure. There's almost nothing macOS specific (there is `-mmacosx-version-min`, which I'm now convinced has almost no effect here), but since it lacks an udev alternative I think it's fair to make a stronger macOS mention.

codingllama committed:
Rename script to build-fido2-macos.sh

codingllama commented:
Thanks, @espadolini.

PTAL @lxea ?

codingllama commented:
Thanks, merging!

codingllama commented:
Actually, it's a bit late here today, so I'll hold this one until tomorrow morning just in case.
